File: 91.cpp
struct Solution {
    int numDecodings(const string &s) {
        if (s[0] == '0'){
            return 0;
        }
        vector<int> memo(s.size() + 1, 0);
        memo[0] = 1;
        memo[1] = 1;
        for (int i = 2; i < s.size() + 1; ++i){
            if (s[i - 1] != '0'){
                memo[i] += memo[i - 1];
            }
            if (s[i - 2] == '1' || (s[i - 2] == '2' && '0' <= s[i - 1] && s[i - 1] <= '6')){
                memo[i] += memo[i - 2];
            }
        }
        return memo.back();
    }
};

File: sol/91/91.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == '0':
            return 0
        memo = [0] * (len(s) + 1)  # [0, e)
        memo[0] = 1
        memo[1] = 1
        for i in range(2, len(s) + 1):
            if s[i - 1] != '0':
                memo[i] += memo[i - 1]
            if s[i - 2] == '1' or (s[i - 2] == '2' and s[i - 1] in '0123456'):
                memo[i] += memo[i - 2]
        return memo[-1]
